service-base/src/main/java/org/eclipse/hono/service/tenant/AbstractTenantAmqpEndpoint.java

```diff
@@ -0,0 +1,310 @@
+
+        final Promise<TenantResult<JsonObject>> getResult = Promise.promise();
+        getService().get(tenantId, span, getResult);
+        return getResult.future().map(tr -> TenantConstants.getAmqpReply(TenantConstants.TENANT_ENDPOINT, tenantId, request, tr));
+    }
+
+    private Future<Message> processGetByCaRequest(final Message request, final String subjectDn,
+            final Span span) {
+
+        try {
+            final X500Principal dn = new X500Principal(subjectDn);
+            log.debug("retrieving tenant [subject DN: {}]", subjectDn);
+            final Promise<TenantResult<JsonObject>> getResult = Promise.promise();
+            getService().get(dn, span, getResult);
+            return getResult.future().map(tr -> {
+                String tenantId = null;
+                if (tr.isOk() && tr.getPayload() != null) {
+                    tenantId = getTypesafeValueForField(String.class, tr.getPayload(),
+                            TenantConstants.FIELD_PAYLOAD_TENANT_ID);
+                    span.setTag(MessageHelper.APP_PROPERTY_TENANT_ID, tenantId);
+                }
+                return TenantConstants.getAmqpReply(TenantConstants.TENANT_ENDPOINT, tenantId, request, tr);
+            });
+        } catch (final IllegalArgumentException e) {
+            TracingHelper.logError(span, "illegal subject DN provided by client: " + subjectDn);
+            // the given subject DN is invalid
+            log.debug("cannot parse subject DN [{}] provided by client", subjectDn);
+            return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST));
+        }
+    }
+
+    /**
+     * Creates a new <em>OpenTracing</em> span for tracing the execution of a tenant service operation.
+     * <p>
+     * The returned span will already contain a tag for the given tenant (if it is not {@code null}).
+     *
+     * @param operationName The operation name that the span should be created for.
+     * @param spanContext Existing span context.
+     * @param tenantId The tenant id.
+     * @return The new {@code Span}.
+     * @throws NullPointerException if operationName is {@code null}.
+     */
+    protected final Span newChildSpan(final String operationName, final SpanContext spanContext, final String tenantId) {
+        Objects.requireNonNull(operationName);
+        // we set the component tag to the class name because we have no access to
+        // the name of the enclosing component we are running in
+        final Tracer.SpanBuilder spanBuilder = TracingHelper.buildChildSpan(tracer, spanContext, operationName)
+                .ignoreActiveSpan()
+                .withTag(Tags.COMPONENT.getKey(), getClass().getSimpleName())
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_SERVER);
+        if (tenantId != null) {
+            spanBuilder.withTag(MessageHelper.APP_PROPERTY_TENANT_ID, tenantId);
+        }
+        return spanBuilder.start();
+    }
```

Comment:
I think that can be moved up or to a Util class so it can be used by the other endpoints

Reply:
Good catch. Meant to have it in the base class, but it slipped.